### PR TITLE
🌱  Update ASG when subnets are changed in AWSMachinePool

### LINKED ISSUE #3315: Update ASG when AWSMachinePool subnets change
We only use subnets fields in AWSMachinePool field during ASG creation and do not update Auto Scaling Group if subnets change.
UpdateASG() should return true if a change is detected in the subnets.
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/6ab95625da2aabb209bc5744883eb4c009cc4fba/exp/controllers/awsmachinepool_controller.go#L534

ASG rollouts all machines and bring up new ones once updated with new `VPCZoneIdentifier`.

/kind feature
/priority important-longterm
/milestone v1.x

### PR DESCRIPTION
**Pull request by DiptoChakrabarty**
**What type of PR is this?**

/kind feature
**What this PR does / why we need it**:

Updates ASG when subnets are changed in AWSMachinePool

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3315 

**Special notes for your reviewer**:
would like to test using github

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

File: pkg/cloud/services/autoscaling/autoscalinggroup.go [hunk 1]
```diff
@@ -44,6 +44,7 @@ func (s *Service) SDKToAutoScalingGroup(v *autoscaling.Group) (*expinfrav1.AutoS
 		MaxSize:           int32(aws.Int64Value(v.MaxSize)),
 		MinSize:           int32(aws.Int64Value(v.MinSize)),
 		CapacityRebalance: aws.BoolValue(v.CapacityRebalance),
+		Subnets:           strings.Split(*v.VPCZoneIdentifier, ", "),
```
Comment (Ankitasw) [inline pkg/cloud/services/autoscaling/autoscalinggroup.go:47]: Do we need the split identifier to be `, ` and not `,` ?

File: exp/controllers/awsmachinepool_controller.go [hunk 2]
```diff
@@ -532,7 +532,9 @@ func asgNeedsUpdates(machinePoolScope *scope.MachinePoolScope, existingASG *expi
 	}
 
 	// todo subnet diff
-
+	if !reflect.DeepEqual(machinePoolScope.AWSMachinePool.Spec.Subnets, existingASG.Subnets) {
```
[the remark this answers is not in the document]
Comment (DiptoChakrabarty) [reply exp/controllers/awsmachinepool_controller.go]: changed it

File: pkg/cloud/services/autoscaling/autoscalinggroup.go (hunk 1 above)
Comment (DiptoChakrabarty) [reply pkg/cloud/services/autoscaling/autoscalinggroup.go:47]: the VPCZoneIdentifier is made like this `aws.String(strings.Join(i.Subnets, ", "))` hence making the subnets by splitting on `, `  else the subnets length increases by one due to extra space

[the remark this answers is not in the document]
Comment (DiptoChakrabarty) [reply pkg/cloud/services/autoscaling/autoscalinggroup.go:47]: added a subnet field

**Comment by Ankitasw** [comment]
@DiptoChakrabarty can you squash your commits?

**Comment by DiptoChakrabarty** [comment]
> can you squash your commits?

done

**Comment by Ankitasw** [comment]
@DiptoChakrabarty can you please run the tests locally and see the issue, it's something to do with these changes only.

**Comment by DiptoChakrabarty** [comment]
@Ankitasw based on the change I should add a  test to check `Subnet != asg.Subnet` right ?

**Comment by Ankitasw** [comment]
> @Ankitasw based on the change I should add a test to check Subnet != asg.Subnet right ?

Yes please.

**Comment by Ankitasw** [comment]
> @Ankitasw based on the change I should add a test to check Subnet != asg.Subnet right ?

Is this change yet to be done?

**Comment by DiptoChakrabarty** [comment]
> Is this change yet to be done?

yes , added the testcase but encountering an error , will push the test soon

**Comment by DiptoChakrabarty** [comment]
the tests are failing because I set the `machinePoolScope.AWSMachinePool.Spec.Subnets` as `[]infrav1.AWSResourceReference{
								{
									ID: pointer.String("subnet1"),
								},
								{
									ID: pointer.String("subnet2"),
								}, 
							},`   and the `existingASG.Subnets` as `[]string{"subnet1", "subnet2"},`  do they have to be set in another format for this case

**Comment by Ankitasw** [comment]
> the tests are failing because I set the machinePoolScope.AWSMachinePool.Spec.Subnets as []infrav1.AWSResourceReference{ { ID: pointer.String("subnet1"), }, { ID: pointer.String("subnet2"), }, }, and the existingASG.Subnets as []string{"subnet1", "subnet2"}, do they have to be set in another format for this case

I am not sure how cmp.Equal will not work with different types of structs. @shivi28 I remember you making a doc on go cmp usage, can you let us know if you came across any such examples as used in this PR?

**Comment by DiptoChakrabarty** [comment]
> I am not sure how cmp.Equal will work with different types of structs. @shivi28 I remember you making a doc on go cmp usage, can you let us know if you came across any such examples as used in this PR?

I can make a go cmp comparer function to compare the `AWSResourceReference.ID` and `AutoScalingGroup.Subnets` however I was suggested to do something like this https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3315#issuecomment-1074293015 just making sure

**Comment by Ankitasw** [comment]
@DiptoChakrabarty were you able to figure out the issue?

**Comment by Ankitasw** [comment]
@DiptoChakrabarty would you like to continue this PR?

**Pull request closed without merging**